code/game/machinery/computer/aifixer.dm

```diff
@@ -48,45 +47,46 @@
 
 		var/laws[0]
 		for(var/datum/ai_law/law in occupant.laws.all_laws())
-			laws.Add(list(list("law" = law.law, "number" = law.get_index())))
-
+			if(law in occupant.laws.ion_laws) // If we're an ion law, give it an ion index code
+				laws.Add(ionnum() + ". " + law.law)
+			else
+				laws.Add(num2text(law.get_index()) + ". " + law.law)
 		data["laws"] = laws
+		data["has_laws"] = length(laws)
 
 	return data
 
-/obj/machinery/computer/aifixer/Topic(href, href_list)
+/obj/machinery/computer/aifixer/tgui_act(action, params)
 	if(..())
-		return 1
-
-	if(href_list["fix"])
-		active = 1
-		while(occupant.health < 100)
-			occupant.adjustOxyLoss(-1, FALSE)
-			occupant.adjustFireLoss(-1, FALSE)
-			occupant.adjustToxLoss(-1, FALSE)
-			occupant.adjustBruteLoss(-1, FALSE)
-			occupant.updatehealth()
-			if(occupant.health >= 0 && occupant.stat == DEAD)
-				occupant.update_revive()
-				occupant.lying = 0
-				update_icon()
-			sleep(10)
-		active = 0
-		add_fingerprint(usr)
-
-	if(href_list["wireless"])
-		var/wireless = text2num(href_list["wireless"])
-		if(wireless == 0 || wireless == 1)
-			occupant.control_disabled = wireless
-
-	if(href_list["radio"])
-		var/radio = text2num(href_list["radio"])
-		if(radio == 0 || radio == 1)
-			occupant.aiRadio.disabledAi = radio
-
-	SSnanoui.update_uis(src)
+		return
+	switch(action)
+		if("fix")
+			active = TRUE
+			INVOKE_ASYNC(src, .proc/fix_ai)
+			add_fingerprint(usr)
+
+		if("wireless")
+			occupant.control_disabled = !occupant.control_disabled
+
+		if("radio")
+			occupant.aiRadio.disabledAi = !occupant.aiRadio.disabledAi
+
 	update_icon()
-	return
+	return TRUE
+
+/obj/machinery/computer/aifixer/proc/fix_ai() // Can we fix it? Probrably.
+	while(occupant.health < 100)
+		occupant.adjustOxyLoss(-1, FALSE)
+		occupant.adjustFireLoss(-1, FALSE)
+		occupant.adjustToxLoss(-1, FALSE)
+		occupant.adjustBruteLoss(-1, FALSE)
+		occupant.updatehealth()
+		if(occupant.health >= 0 && occupant.stat == DEAD)
+			occupant.update_revive()
+			occupant.lying = FALSE
+			update_icon()
+		sleep(10)
+	active = FALSE
```

Comment:
Add a check here to make sure you cant spam this to rapidly fix the AI, or consume limitless CPU cycles

Reply:
handled by disabling the fix button when fixing
`disabled={integrityFull || data.active}`